Refactor loss

marcociccone opened the pull request:
This is a PR to start a refactoring of the main loop and make it more modular.
The final goal is to have a simpler structure where for every project one should just write his own:
* build_model
* buil_loss
* validation_fn

Optimistically we can have callback functions a la keras :)

marcociccone committed:
Refactoring:
- Add build_loss: everyone should write his own loss function
- Input and outputs are now contained in a dictionary to be easily
passed between functions

marcociccone committed:
Change dictionary names, Clean old code

marcociccone committed:
Restore compatibility with dataset_loader

pep8speaks commented:
Hello @marcociccone! Thanks for updating the PR.









 - In the file [`main_loop_tf/main.py`](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/main.py), following are the PEP8 issues :

> [Line 5:1](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/main.py#L5): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 33:1](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/main.py#L33): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'

 - In the file [`main_loop_tf/utils.py`](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/utils.py), following are the PEP8 issues :

> [Line 257:80](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/utils.py#L257): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 267:80](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/utils.py#L267): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

 - In the file [`main_loop_tf/validate.py`](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/validate.py), following are the PEP8 issues :

> [Line 3:1](https://github.com/fvisin/main_loop_tf/blob/0903a1534296b083e3c3414a5252b462f305652d/main_loop_tf/validate.py#L3): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'



##### Comment last updated on September 22, 2017 at 20:41 Hours UTC

fvisin commented on main_loop_tf/main.py:
```diff
@@ -47,18 +43,17 @@
                    'middle frame prediction')
-gflags.DEFINE_bool('return_middle_frame_only', False, 'If True, return '
-                   'the middle frame segmentation mask only for each sequence')
+gflags.DEFINE_bool('return_middle_frame_only', True, '')
```
Why did you delete the description? I guess you modified the behaviour of the flag, can you update it with the new one?

fvisin commented on main_loop_tf/main.py:
```diff
@@ -90,7 +85,7 @@ def __parse_config(argv=None):
                     'group_summaries', 'help', 'hyperparams_summaries',
                     'max_epochs', 'min_epochs', 'model_name', 'nthreads',
-                    'patience', 'return_middle_frame_only', 'restore_model',
+                    'patience', 'restore_model',
```
Ok, so, probably you intended to remove the flag before? Can you make these two modifications consistent?

fvisin commented on main_loop_tf/main.py:
```diff
@@ -142,12 +137,14 @@ def __parse_config(argv=None):
     # Dataset
     try:
-        Dataset = getattr(dataset_loaders, cfg.dataset)
+        Dataset = getattr(dataset_loaders, cfg.dataset + 'Dataset')
```
I'd rather go with a third `try/except` statement so that this works also if the dataset name specified in the flag is complete.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -161,24 +158,11 @@ def __parse_config(argv=None):
-            cfg.val_input_shape[-1] = 6
-
-        if cfg.crop_size:
-            cfg.input_shape[2:4] = cfg.crop_size
```
I am not sure that the new way you deal with the `input_shape` is correct, as I commented [here](https://github.com/fvisin/main_loop_tf/pull/19#commitcomment-23923388). Please let me know your thoughts on what I wrote in the comment.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -198,32 +182,61 @@ def __parse_config(argv=None):
+        cfg.val_input_shape = [None] + list(
+            valid_temp.next()['data'].shape[1:])
+        if cfg.crop_size:
+            cfg.input_shape[1:3] = cfg.crop_size
```
To summarize my thoughts again, you don't want the input/val_input shapes to be tied to the shape of the first element returned by the dataset. If the elements of the dataset don't have a fixed shape this will not work, or worse, will not work as intended.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -301,6 +314,7 @@ def __run(build_model):
         val_placeholders = [val_inputs_per_gpu, labels_per_gpu, num_splits,
                             num_batches]
 
+        # TODO: move LR schedule inside Object Optimizer, to be created
```
Add this as a TODO in the first comment of this PR please. 
If you add a list of checkboxes github will automatically show the progress of the PR! ;)

fvisin commented on main_loop_tf/main.py:
```diff
@@ -334,6 +348,10 @@ def __run(build_model):
+
+        elif cfg.lr_decay == 'STN':
+            epoch = tf.cast(cfg.global_step / cfg.decay_steps, tf.int32)
+            lr = cfg.lr * tf.pow(0.5, tf.cast(epoch / 50, cfg._FLOATX))
```
This should probably be moved in our repo once this PR is final and merged.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -504,35 +523,54 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+            # TODO: create **loss_params to  be defined in model repo
+            loss_dict = build_loss(dev_labels, model_out_dict, loss_fn,
+                                   l2_reg=weight_decay,
+                                   inputs=dev_inputs)
```
Why do you need `dev_inputs` here? Shouldn't you use one of the `model_out_dict` outputs (I'd say model_out_dict['out_act'], but it's probably best to pass the whole dict directly)?

fvisin replied:
Also I'd like your opinion on this: should we only pass `weight_decay` as you are doing or rather the whole `cfg` dictionary? The second option would allow for any number of parameters to be set and retrieved in the loss by the user.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -504,35 +523,54 @@ def build_graph(placeholders, input_shape, build_model, which_set):
             with tf.name_scope(None):
                 with tf.name_scope(dev_set_str + '_stats') as dev_set_scope:
-                    tf.summary.scalar('Loss', loss, summaries)
+                    tf.summary.scalar('Loss', loss_dict['loss'], summaries)
```
Should we have a summary for `loss_dict['components']` as well? If so, should we cycle over the `loss_dict` keys and create a summary for each one of them?

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
-    preds_flat = tf.reshape(preds, [-1])
-    tf.summary.scalar('control_flow/batch_size_' + which_set,
-                      tf.shape(preds)[0], summaries)
+    # Towers contain dictionary
```
Delete or improve the comment? :)

fvisin replied:
It's actually useful to add a comment here, as it was not obvious to me at first that each item in `tower_out_dict` is a list.

I suggest:
Each item in \`tower_out_dict\` is a list of tensors, which we want to convert to a tensor (by concatenation) in order to then retain only the actual number of batches we care about (in case the input batch is shorter than the placeholder size)

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
-        mask = tf.cast(tf.less_equal(labels, nclasses), tf.int32)
-    m_iou, per_class_iou, cm_update_op, reset_cm_op = compute_mean_iou(
-        labels, preds_flat, nclasses, mask)
+    # (equivalent to labels[:np.prod(out_dict.shape)])
```
out_dict.shape --> out_dict['pred'].shape

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+    # Towers contain dictionary
+    out_dict = {}
+    for k, v in tower_out_dict.iteritems():
+        # Convert from list of tensors to catenated tensor
```
You can remove this comment

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+    out_dict = {}
+    for k, v in tower_out_dict.iteritems():
+        # Convert from list of tensors to catenated tensor
+        out_dict[k] = tf.concat(tower_out_dict[k], axis=0,
```
`tower_out_dict[k]` is `v` :)

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+        # Convert from list of tensors to catenated tensor
+        out_dict[k] = tf.concat(tower_out_dict[k], axis=0,
+                                name='concat_%s' % k)
+        out_dict[k] = out_dict[k][:num_batches]
```
It should work also inplace, which would reduce the proliferation of variables:
``` python
tower_out_dict[k] = tf.concat(v, axis=0, name='concat_%s' % k)[:num_batches]
```

fvisin commented on main_loop_tf/main.py:
```diff
@@ -504,35 +523,54 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+            assert 'out_act' in model_out_dict, """Your model function should
+                return a dictionary with attribute 'out_act'!"""
+            assert 'pred' in model_out_dict, """Your model function should
+                return a dictionary with at least attribute 'pred'!"""
```
For the three of them, it would be nice to explain what is expected to be in each variable (e.g., the activation before/after the activation function, the prediction after the argmax)

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+    # (equivalent to labels[:np.prod(out_dict.shape)])
+    labels = labels[:tf.shape(tf.reshape(out_dict['pred'], [-1]))[0]]
+
+    # TODO: add metrics callback
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
     tf.summary.scalar(scope_str % 'Mean_tower_loss', avg_tower_loss, summaries)
 
+    # Compute the average of the loss per each component
+    # (Just for visualization purpose)
```
purposes

fvisin commented on main_loop_tf/main.py:
```diff
@@ -674,10 +737,18 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+            metrics_out.append(m_iou, per_class_iou)
+
+        outs = {}
+        outs.update(out_dict)
```
Why not using `out_dict` directly?
Furthermore, since it's a dictionary and the specific keys can be accessed later on in the code, I suggest to use it for training as well. This way you can update it with `avg_tower_loss` before the if/else branches and you just need to add the `train_ops` in one case and the `metrics_out` and the `cm_update_op` in the other.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -674,10 +737,18 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+        outs.update(out_dict)
+        outs.update({'metrics_out': metrics_out,
+                     'avg_tower_loss': avg_tower_loss})
+        if cm_update_op is not None:
```
You can remove the if. When it's used later on I think it's cleaner to check for `outs['cm_update_op'] is None` rather than `hasattr(outs, 'cm_update_op')`

fvisin replied:
BTW, you already do it for `reset_cm_op`.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -754,7 +824,7 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_ops,
 
     while not sv.should_stop():
-        epoch_id = cum_iter // train.nbatches
+        epoch_id = (cum_iter+1) // train.nbatches
```
I don't think it's correct. The id starts from 0 and we already add 1 when we visualize.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,40 +656,65 @@ def build_graph(placeholders, input_shape, build_model, which_set):
+            labels, pred_flat, nclasses, mask)
+    else:
+        cm_update_op = None
+        reset_cm_op = None
```
reset_cm_op --> cm_reset_op

fvisin commented on main_loop_tf/main.py:
```diff
@@ -712,7 +783,6 @@ def build_graph(placeholders, input_shape, build_model, which_set):
     for _, s in enumerate(summaries[::-1]):
         summary_ops.append(tf.summary.merge(tf.get_collection_ref(key=s)))
-
     return outs, summary_ops, reset_cm_op
```
Given that you are returning a dictionary, I suggest to put `summary_ops` and `reset_cm_op` in it as well.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -863,6 +939,10 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_ops,
+
+            # TODO: ok so far... but validate should be a wrapper too
+            # because it is task dependent and you can have different
+            # metrics or visualizations
```
Ditto

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -98,26 +100,33 @@ def validate(placeholders,
+                    # reset_states(model, x_batch.shape)
+                prev_subset = subset
+
+        if cfg.seq_length and y_batch.shape[1] > 1:
```
Why don't you use `cfg.seq_length` here? There should never be the case that a sequence can have different length than `seq_length`.

fvisin commented on main_loop_tf/validate.py:
Shouldn't this return a dictionary as in the training case? If no, can we do it?

fvisin commented on main_loop_tf/validate.py:
Ditto

fvisin commented on main_loop_tf/validate.py:
I believe mIoU should be set to 0 also in the other branch (i.e., once for both branches, before L193)

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -280,6 +299,8 @@ def write_summary(labs, val):
     import matplotlib as mpl
     import seaborn as sns
+    from utils import flowToColor
+    import cv2
```
I don't like cv2 to be a mandatory requirement for main_loop_tf. Move the import where you use it please, and try not to whenever is possible :)

fvisin reviewed (changes requested):
I did not review the validation and the visualization functions in detail, as they are a work in progress. 

The rest looks very good to me, I only have two high level requests:
1) remove the "dict" from the variable names. I don't think it's necessary and I think that shorter variable names would be better
2) once the PR is final and ready to merge, please check that the main_loop uses the same variable names that are used for the outputs of build_graph

fvisin committed:
Create Experiment object

fvisin committed:
Simplify the code by using the Experiment object

- Try to encapsulate every information about the current epoch /
    iteration in the object (something might be missing)
- The validate function should be modified so that
    - it becomes a function of the Experiment class
    - it gathers its inputs from the object's state rather than the
    parameters of the function
- There are a few things I noticed during this refactoring:
    - `metrics_out` is weirdly never used
    - `cm_update_ops` is weirdly never used

marcociccone committed:
Decoupling methods:

* make Experiment class Abstract
* build_model, build_loss must be implemented
* remove validation: should be task dependent
* [WIP] decouple metrics: should be passed as list
* remove old comments, old parameters

fvisin committed:
Fix input ordering (for good)

* Tentative fix in a373001 was not right. This commit builds on that to
  fix the issue with the weird ordering of the outputs of the model for
  good.
* Compute the incremental summaries operations without resorting to the
  inverted loop in the devices, which was difficult to understand
* Improve explanation of what is going on in the code

fvisin committed:
Minor cosmetic changes

fvisin committed:
loss --> optimization

fvisin committed:
Split graph creation function

* Split the graph creation into two class methods, a per-device one and
  a general one that calls the per-device multiple times.

fvisin committed:
Create an Optimizer object

fvisin committed:
Create run and validate methods

* Move the graph initialization code in a ad-hoc class function
* Create a run and a validation class method

fvisin committed:
Cosmetic changes

fvisin committed:
Move gradient processing inside Optimizer

* BROKEN: the import in main.py should be updated

fvisin committed:
Add DistributedOptimizer

* Move the gradient computation and processing (noise, multiplier, ..)
  and related summaries inside the Optimizer object.
* Make the Optimizer object stateful so as to keep trace of the gradient
  of the devices already processed

fvisin committed:
Move average grad computation in Optimizer

* Move the computation of the average gradients and of the per-component mean gradients in the Optimizer object

fvisin committed:
Expose Experiment from main_loop_tf directly

fvisin committed:
Fix signature in abstract methods

fvisin committed:
Fix Experiment object

fvisin committed:
model_out --> model_outs

fvisin committed:
Fix object creation order

fvisin committed:
Fix DistributedOptimizer dynamic subclassing

fvisin committed:
Fix unknown grad_noise error management

fvisin committed:
loss_dict --> loss_outs

fvisin committed:
Fix Optimizer

* Check 1:1 consistency with previous code
* Improve names and enforce consistency throughout the code
* Push back in the main the loss / comp loss summaries

fvisin committed:
Remove the optimizer from the Experiment state

* The optimizer object is now stateful and contains the per-device
  information relative to a specific graph (i.e., train or validation
  graph). We don't want to have a single Optimizer object for all of
  them, as those graphs are separate.

fvisin committed:
Use the same names in graph creation and run

fvisin committed:
Add main_loop callbacks

fvisin committed:
Fix validation

* Let the user define the validation method
* Run validation at the end of the epoch only if specified

fvisin committed:
Fix callbacks

fvisin committed:
Fix avg_loss missing

fvisin committed:
Fix cum_iter not increasing

fvisin committed:
Make graph output more flexible

* __build_device_graph returns a dictionary
* the graph outputs are saved in the Experiment state as a dictionary
* add a compute_metrics function that can be defined by the user to
  compute some metric in the graph
* rename some variable to be more explicit

fvisin committed:
_build_device_graph__ --> _build_device_graph

fvisin committed:
Add signature of the functions to be implemented by user

* Not really a great solution but I couldn't cum up with anything
  better. For the time being it will do.

fvisin committed:
compute_metrics --> metrics_graph_fn

fvisin committed:
Add evaluate function, improve validation

fvisin committed:
Remove val_inputs_per_gpu argument as is in the state

fvisin committed:
Fix do not create multiple global_steps

fvisin committed:
Move loss_fn flag into model

fvisin committed:
Cosmetic changes

fvisin committed:
cum_iter --> gstep_val + reorder step computation

fvisin committed:
Move main loop logic into callbacks

fvisin committed:
Move avg loss computation outside of the optimizer

fvisin committed:
distributed_minimize --> minimize

* Extend the interface of the original minimize() function rather than
  creating a new one

fvisin committed:
Clearly separate loss and gradient computation

fvisin committed:
Move avg loss computation outside of the optimizer

* Fix dev_losses overwritten when several valid sets
* Do not store dev_losses in the state, so that they do not exist over
  multiple calls of build_device_graph
* Move the mean computation outside of the optimizer to clearly separate
  gradients and losses

fvisin committed:
evaluate --> validate

fvisin committed:
__build_graph__ --> __build_graph

fvisin committed:
Allow user-defined placeholders

* Collect the placeholders via a function, so that the user can override
  it and add his own placeholders
* Restructure the code to work with lists of dictionaries, one for each
  device
* Restructure the main loop accordingly to create lists of dictionaries
  of inputs

fvisin committed:
Fix gradient noise based on prev err

* Update the prev loss *after* the new loss is computed
* Pass None when the noise should not be added to the gradient

fvisin committed:
Allow the user to specify which vars to optimize

fvisin committed:
Allow the user to specify some extra loss function params

fvisin committed:
Move reconvnets code into run

fvisin committed:
Add todos

fvisin committed:
Add option to exclude extra flags from hash

fvisin committed:
Add process_cfg_flags function

fvisin committed:
Add smart-cropping

fvisin committed:
docstring_get_grad_desc_var_list

fvisin committed:
Add callbacks for graph creation

* Allow the user to:
    * Process the model output for visualization
    * Add user-defined per-device summaries
    * Add user-defined global summaries
    * Add user-defined metrics to the graph

fvisin committed:
Fix bug in summaries names scopes

fvisin committed:
Add feed_dict to state

fvisin committed:
Do not override return_middle_frame_only in validation

* Do not set the flag to False by default in validation. Let the user
  choose

fvisin committed:
Do not ignore user model modifications
